kmod/src/kernelcompat.h

```diff
@@ -174,33 +174,32 @@ do {					\
 
 #ifndef KC_SHRINKER_SHRINK
 #define KC_DEFINE_SHRINKER(name) struct shrinker name
-#define KC_INIT_SHRINKER_FUNCS(type, name, shrink, count, scan) do {	\
-	__typeof__(shrink) _shrink = (shrink);				\
-	_shrink->count_objects = count;					\
-	_shrink->scan_objects = scan;					\
+#define KC_INIT_SHRINKER_FUNCS(type, name, shrinkfn, countfn, scanfn) do {	\
+	__typeof__(shrinkfn) _shrink = (shrinkfn);				\
+	_shrink->count_objects = (countfn);					\
+	_shrink->scan_objects = (scanfn);					\
 } while (0)
 #else
 #include <linux/shrinker.h>
+int kc_shrink_wrapper(struct shrinker *shrink, struct shrink_control *sc);
 struct kc_shrinker_funcs {
 	unsigned long (*count_objects)(struct shrinker *, struct shrink_control *sc);
 	unsigned long (*scan_objects)(struct shrinker *, struct shrink_control *sc);
 };
 /* using adjacent member of an unnamed struct */
-#define KC_DEFINE_SHRINKER(name)				\
+#define KC_DEFINE_SHRINKER(name) struct				\
 	{							\
 		struct kc_shrinker_funcs shrinker_funcs;	\
-		struct shinker name;				\
+		struct shrinker (name);				\
 	}
-#define KC_SHRINKER_FUNCS(shrinker) \
-	((void *)((long)(shrink) - sizeof(struct kc_shrinker_funcs)))
-#define KC_INIT_SHRINKER_FUNCS(type, name, shrink, count, scan) do {				\
-	BUILD_BUG_ON(offsetof(cont, shrink_funcs) + sizeof(struct kc_shrinker_funcs)) !=	\
```

Comment:
Hmm, removing the BUILD_BUG_ON removes the protection from padding and struct re-ordering of the unnamed struct.  Maybe the better fix is to use a real named struct (maybe kc_shrinker_container) so we can then use container_of in _SHRINKER_FUNCS instead of the raw pointer math.

Reply:
:)

The biggest problem I had here is that I just couldn't come up with a way to make this BUILD_BUG_ON() actually compile. Note the typo (`cont`)  but even that isn't the problem. I end up with this wonderful error message salad:

```
rc/block.c:1296:28: error: expected specifier-qualifier-list before ‘block_count_objects’
           &binf->shrinker, block_count_objects, block_scan_objects);
                            ^
include/linux/compiler.h:434:19: note: in definition of macro ‘__compiletime_assert’
   bool __cond = !(condition);    \
                   ^
include/linux/compiler.h:454:2: note: in expansion of macro ‘_compiletime_assert’
  _compiletime_assert(condition, msg, __compiletime_assert_, __LINE__)
  ^
include/linux/bug.h:53:37: note: in expansion of macro ‘compiletime_assert’
 #define BUILD_BUG_ON_MSG(cond, msg) compiletime_assert(!(cond), msg)
                                     ^
include/linux/bug.h:77:2: note: in expansion of macro ‘BUILD_BUG_ON_MSG’
  BUILD_BUG_ON_MSG(condition, "BUILD_BUG_ON failed: " #condition)
  ^
/home/auke/scoutfs/kmod/src/kernelcompat.h:228:2: note: in expansion of macro ‘BUILD_BUG_ON’
  BUILD_BUG_ON(offsetof(countfn, shrinker_funcs) + sizeof(struct kc_shrinker_funcs) != \
  ^
include/linux/stddef.h:16:32: note: in expansion of macro ‘__compiler_offsetof’
 #define offsetof(TYPE, MEMBER) __compiler_offsetof(TYPE, MEMBER)
                                ^
/home/auke/scoutfs/kmod/src/kernelcompat.h:228:15: note: in expansion of macro ‘offsetof’
  BUILD_BUG_ON(offsetof(countfn, shrinker_funcs) + sizeof(struct kc_shrinker_funcs) != \
               ^
/home/auke/scoutfs/kmod/src/block.c:1295:2: note: in expansion of macro ‘KC_INIT_SHRINKER_FUNCS’
  KC_INIT_SHRINKER_FUNCS(struct block_info, shrinker,
  ^
```

I do see what you intend to do here... I'm gonna skip over this for now though.

Reply:
Yeah, the unnamed struct is gumming things up.  Probably just better to make a _container struct and then use container_of instead of the sketchy manual pointer math..